corpus_studies/solidity_state_study/slither_plugins/ContractStateDetector.py

```diff
@@ -117,43 +129,55 @@ def fields_in_getter_node(node: Node) -> Set[Variable]:
     # Checks if the node makes a stateful check
     # Returns a set of state variables used, or an empty set if none are found
     # constant_parameters are the variables (in modifiers) which we know to be constant values.
-    def statevars_in_node(self, node: Node, func: Function, constant_parameters: Set[Variable] = None) -> Set[StateVariable]:
+    def states_in_node(self, node: Node, func: Function, constant_parameters: Set[Variable] = None) -> Optional[StateEvidence]:
         if constant_parameters is None:
             constant_parameters = set()
         # Return a set of all (nonconstant) state vars that the variable is dependent on
-        def state_vars_used(var: Variable) -> Set[StateVariable]:
+        def state_vars_used(var: Variable, seen = None) -> Set[StateVariable]:
+            if seen is None: seen = set()
             if var in self.state_vars:
                 if var in self.nonconstant_vars:
                     return {var}
                 else:
                     return set()
             else:
-                # If var is dependent on itself, remove it to stop infinite recursion
-                next_vars = get_dependencies(var, func) - {var}
-                return set().union(*map(state_vars_used, next_vars))
+                # If var has been seen before, stop to prevent infinite recursion
+                if var in seen:
+                    return set()
+                seen.add(var)
+                next_vars = get_dependencies(var, func)
+                return set().union(*map(lambda v: state_vars_used(v, seen), next_vars))
         # Check if a variable is dependent only on state variables
-        def is_dependent_on_state_vars(var: Variable):
+        def is_dependent_on_state_vars(var: Variable, seen = None):
```

Comment:
This is a Python idiom, where default parameters are set to None because default parameters are only initialized once when the function is declared, rather than when the function is called. This can be an issue for instance when using mutable sets.